// tdesign-vue-next/es/tree/adapt.mjs
/**
 * tdesign v1.10.2
 * (c) 2024 tdesign
 * @license MIT
 */

import * as Vue from 'vue';
export { TransitionGroup, computed, defineComponent, onMounted, reactive, ref, toRefs, watch } from 'vue';
import withInstall$1 from '../utils/withInstall.mjs';
import useVModel$1 from '../hooks/useVModel.mjs';
import useDefaultValue from '../hooks/useDefaultValue.mjs';
export { CaretRightSmallIcon as TdCaretRightSmallIcon } from 'tdesign-icons-vue-next';
export { Checkbox as TCheckBox } from '../checkbox/index.mjs';
export { Loading as TLoading } from '../loading/index.mjs';
export { usePrefixClass } from '../hooks/useConfig.mjs';
export { useGlobalIcon } from '../hooks/useGlobalIcon.mjs';
export { default as useLazyLoad } from '../hooks/useLazyLoad.mjs';
export { default as useVirtualScroll } from '../hooks/useVirtualScrollNew.mjs';
export { TreeNode, privateKey } from '../_common/js/tree/tree-node.mjs';
export { useConfig } from '../config-provider/useConfig.mjs';
import '../_chunks/dep-b88cf96a.mjs';
import '../_chunks/dep-95eb038b.mjs';
import '../_chunks/dep-82139152.mjs';
import '../_chunks/dep-c3300d46.mjs';
import '../_chunks/dep-19284555.mjs';
import '../_chunks/dep-670ed189.mjs';
import '../_chunks/dep-0e4e2bcf.mjs';
import '../_chunks/dep-9a145a1d.mjs';
import '../_chunks/dep-b02becad.mjs';
import '../_chunks/dep-5b72198e.mjs';
import '../checkbox/checkbox.mjs';
import '../_chunks/dep-97ca0f39.mjs';
import '../_chunks/dep-9213781b.mjs';
import '../_chunks/dep-5597cc76.mjs';
import '../checkbox/props.mjs';
import '../hooks/useRipple.mjs';
import '../hooks/useKeepAnimation.mjs';
import '../_chunks/dep-d22c9e6d.mjs';
import '../_chunks/dep-18e89938.mjs';
import '../_chunks/dep-50ae0ec6.mjs';
import '../_chunks/dep-c32ef8ef.mjs';
import '../_chunks/dep-6f372498.mjs';
import '../_chunks/dep-48219979.mjs';
import '../_chunks/dep-2b538834.mjs';
import '../_chunks/dep-eae00e6e.mjs';
import '../_chunks/dep-40d5a6ec.mjs';
import '../_chunks/dep-564c533a.mjs';
import '../_chunks/dep-1b4b86fa.mjs';
import '../_chunks/dep-8eb27cbe.mjs';
import '../_chunks/dep-257497ac.mjs';
import '../_chunks/dep-f5c04f3c.mjs';
import '../_chunks/dep-6623c796.mjs';
import '../_chunks/dep-70cd6787.mjs';
import '../_chunks/dep-a27f8ab7.mjs';
import '../_chunks/dep-39e58764.mjs';
import '../_chunks/dep-a2bb880f.mjs';
import '../_chunks/dep-bf0849ad.mjs';
import '../_chunks/dep-4ef46503.mjs';
import '../_common/js/global-config/default-config.mjs';
import '../_common/js/global-config/locale/zh_CN.mjs';
import '../_chunks/dep-257af9e6.mjs';
import '../_chunks/dep-c34f015e.mjs';
import '../_chunks/dep-7e9fb780.mjs';
import '../utils/set-style.mjs';
import '../hooks/tnode.mjs';
import '../_chunks/dep-8d9b7c94.mjs';
import '../_chunks/dep-399b9c3e.mjs';
import '../_chunks/dep-684927fd.mjs';
import '../utils/render-tnode.mjs';
import '../_chunks/dep-dfc325a2.mjs';
import '../_chunks/dep-31a95ba9.mjs';
import '../_chunks/dep-c4e7b9d7.mjs';
import '../_chunks/dep-6e744142.mjs';
import '../checkbox/constants.mjs';
import '../checkbox/hooks/useCheckboxLazyLoad.mjs';
import '../_common/js/utils/observe.mjs';
import '../checkbox/hooks/useKeyboardEvent.mjs';
import '../_common/js/common.mjs';
import '../hooks/useDisabled.mjs';
import '../_chunks/dep-47d8cf00.mjs';
import '../hooks/useReadonly.mjs';
import '../_chunks/dep-95651bdf.mjs';
import '../_chunks/dep-37133d6d.mjs';
import '../_chunks/dep-3f5bea17.mjs';
import '../config-provider/type.mjs';
import '../checkbox/group.mjs';
import '../_chunks/dep-5ce31aa3.mjs';
import '../_chunks/dep-272a424d.mjs';
import '../_chunks/dep-644e70b3.mjs';
import '../_chunks/dep-0272e73d.mjs';
import '../_chunks/dep-ab0f22e3.mjs';
import '../checkbox/checkbox-group-props.mjs';
import '../hooks/slot.mjs';
import './style/css.mjs';
import '../checkbox/type.mjs';
import '../_chunks/dep-a210005e.mjs';
import '../_chunks/dep-ff970c88.mjs';
import '../_chunks/dep-945356ed.mjs';
import '../_chunks/dep-e9c00cc1.mjs';
import '../_chunks/dep-688d94c5.mjs';
import '../_chunks/dep-a9ad4d99.mjs';
import '../_chunks/dep-0554a988.mjs';
import '../loading/plugin.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/set-style.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-209a9bdd.mjs';
import '../_chunks/dep-c3ca6d75.mjs';
import '../_chunks/dep-d45d8d10.mjs';
import '../utils/dom.mjs';
import '../utils/easing.mjs';
import '../loading/props.mjs';
import '../hooks/useTeleport.mjs';
import '../loading/type.mjs';
import '../hooks/useResizeObserver.mjs';
import '../_chunks/dep-cf4a2c01.mjs';
import '../_chunks/dep-8b2e46b6.mjs';
import '../_chunks/dep-6eed8561.mjs';
import '../_chunks/dep-c7f689d1.mjs';
import '../_chunks/dep-2c57cd88.mjs';
import '../_chunks/dep-1f263304.mjs';
import '../_chunks/dep-c0dc4756.mjs';
import '../_common/js/tree/tree-node-model.mjs';
import '../_chunks/dep-3aafefae.mjs';
import '../_chunks/dep-385f320b.mjs';
import '../_chunks/dep-d53d13e3.mjs';
import '../_common/js/log/log.mjs';
import '../_common/js/log/index.mjs';

var onDrag = {
  "default": void 0
};
var isVueNext = true;
var TreeItemDefinition = {
  name: "TTreeNode",
  inject: {
    onDrag: onDrag
  }
};
function useRipple(el) {}
function withInstall(construct) {
  return withInstall$1(construct);
}
function getCreateElement(h) {
  if (h) {}
  return Vue.h;
}
function getScopedSlots(instance) {
  return instance.$slots;
}
function useVModel(props, refsProps) {
  var propName = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "value";
  var defaultPropName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "defaultValue";
  var eventPropName = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : "onChange";
  var eventName = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : "change";
  if (eventName) {}
  var modelValue = refsProps.modelValue;
  var vm;
  if (propName === "value") {
    vm = useVModel$1(refsProps[propName], modelValue, props[defaultPropName], props[eventPropName], propName);
  } else {
    vm = useDefaultValue(refsProps[propName], props[defaultPropName], props[eventPropName], propName);
  }
  return vm;
}

export { TreeItemDefinition, getCreateElement, getScopedSlots, isVueNext, useRipple, useVModel, withInstall };
