// tdesign-vue-next/es/color-picker/panel/format/index.mjs
/**
 * tdesign v1.10.2
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _toConsumableArray } from '../../../_chunks/dep-5ce31aa3.mjs';
import { _ as _defineProperty } from '../../../_chunks/dep-5597cc76.mjs';
import { isVNode, defineComponent, ref, watch, createVNode } from 'vue';
import '../../../_chunks/dep-95eb038b.mjs';
import { _ as _createCompounder } from '../../../_chunks/dep-82139152.mjs';
import props from '../../props.mjs';
import { FORMATS } from '../../const.mjs';
import { Select, Option } from '../../../select/index.mjs';
import FormatInputs from './inputs.mjs';
import { useBaseClassName } from '../../hooks.mjs';
import '../../../_chunks/dep-9213781b.mjs';
import '../../../_chunks/dep-670ed189.mjs';
import '../../../_chunks/dep-c3300d46.mjs';
import '../../../_chunks/dep-19284555.mjs';
import '../../../_chunks/dep-0e4e2bcf.mjs';
import '../../../_chunks/dep-9a145a1d.mjs';
import '../../../_chunks/dep-b02becad.mjs';
import '../../../_chunks/dep-5b72198e.mjs';
import '../../../_common/js/color-picker/constants.mjs';
import '../../../select/select.mjs';
import '../../../_chunks/dep-209a9bdd.mjs';
import '../../../_chunks/dep-97ca0f39.mjs';
import '../../../_chunks/dep-c7559ea1.mjs';
import '../../../_chunks/dep-688d94c5.mjs';
import '../../../_chunks/dep-50ae0ec6.mjs';
import '../../../_chunks/dep-c32ef8ef.mjs';
import '../../../_chunks/dep-6f372498.mjs';
import '../../../_chunks/dep-48219979.mjs';
import '../../../_chunks/dep-2b538834.mjs';
import '../../../_chunks/dep-eae00e6e.mjs';
import '../../../_chunks/dep-40d5a6ec.mjs';
import '../../../_chunks/dep-6ad78111.mjs';
import '../../../_chunks/dep-257497ac.mjs';
import '../../../_chunks/dep-f5c04f3c.mjs';
import '../../../_chunks/dep-70cd6787.mjs';
import '../../../_chunks/dep-a9ad4d99.mjs';
import '../../../_chunks/dep-6623c796.mjs';
import '../../../_chunks/dep-1b4b86fa.mjs';
import '../../../_chunks/dep-d53d13e3.mjs';
import '../../../_chunks/dep-3f5bea17.mjs';
import '../../../_chunks/dep-a2bb880f.mjs';
import '../../../_chunks/dep-39e58764.mjs';
import '../../../_chunks/dep-8b2e46b6.mjs';
import '../../../_chunks/dep-6eed8561.mjs';
import '../../../_chunks/dep-95651bdf.mjs';
import '../../../_chunks/dep-37133d6d.mjs';
import '../../../_chunks/dep-18e89938.mjs';
import '../../../_chunks/dep-564c533a.mjs';
import '../../../_chunks/dep-8eb27cbe.mjs';
import '../../../_chunks/dep-31a95ba9.mjs';
import '../../../_chunks/dep-c4e7b9d7.mjs';
import '../../../_chunks/dep-e9c00cc1.mjs';
import '../../../_chunks/dep-272a424d.mjs';
import '../../../_chunks/dep-644e70b3.mjs';
import '../../../_chunks/dep-0272e73d.mjs';
import '../../../_chunks/dep-a27f8ab7.mjs';
import '../../../common-components/fake-arrow.mjs';
import '../../../hooks/useConfig.mjs';
import '../../../config-provider/useConfig.mjs';
import '../../../_chunks/dep-6e744142.mjs';
import '../../../_chunks/dep-d22c9e6d.mjs';
import '../../../_chunks/dep-bf0849ad.mjs';
import '../../../_chunks/dep-4ef46503.mjs';
import '../../../_common/js/global-config/default-config.mjs';
import '../../../_common/js/global-config/locale/zh_CN.mjs';
import '../../../_chunks/dep-257af9e6.mjs';
import '../../../_chunks/dep-c34f015e.mjs';
import '../../../_chunks/dep-7e9fb780.mjs';
import '../../../config-provider/type.mjs';
import '../../../select-input/index.mjs';
import '../../../select-input/select-input.mjs';
import '../../../popup/index.mjs';
import '../../../popup/popup.mjs';
import '@popperjs/core';
import '../../../hooks/index.mjs';
import '../../../hooks/icon.mjs';
import '../../../hooks/slot.mjs';
import '../../../utils/render-tnode.mjs';
import '../../../_chunks/dep-dfc325a2.mjs';
import '../../../_chunks/dep-8d9b7c94.mjs';
import '../../../_chunks/dep-399b9c3e.mjs';
import '../../../_chunks/dep-684927fd.mjs';
import '../../../_chunks/dep-b88cf96a.mjs';
import '../../../hooks/tnode.mjs';
import '../../../hooks/useCommonClassName.mjs';
import '../../../hooks/useDefaultValue.mjs';
import '../../../hooks/useDestroyOnClose.mjs';
import '../../../hooks/useKeepAnimation.mjs';
import '../../../hooks/useRipple.mjs';
import '../../../utils/set-style.mjs';
import '../../../hooks/useVirtualScroll.mjs';
import '../../../hooks/useVModel.mjs';
import '../../../hooks/useImagePreviewUrl.mjs';
import '../../../_common/js/upload/utils.mjs';
import '../../../_common/js/log/log.mjs';
import '../../../utils/dom.mjs';
import '../../../utils/easing.mjs';
import '../../../_common/js/utils/set-style.mjs';
import '../../../popup/container.mjs';
import '../../../popup/props.mjs';
import '../../../hooks/useResizeObserver.mjs';
import '../../../utils/withInstall.mjs';
import './style/css.mjs';
import '../../../popup/type.mjs';
import '../../../select-input/props.mjs';
import '../../../select-input/useSingle.mjs';
import '../../../input/index.mjs';
import '../../../input/input.mjs';
import 'tdesign-icons-vue-next';
import '../../../input/props.mjs';
import '../../../hooks/useDisabled.mjs';
import '../../../_chunks/dep-47d8cf00.mjs';
import '../../../hooks/useReadonly.mjs';
import '../../../hooks/useGlobalIcon.mjs';
import '../../../input/useInput.mjs';
import '../../../form/const.mjs';
import '../../../input/useLengthLimit.mjs';
import '../../../_common/js/log/index.mjs';
import '../../../_common/js/utils/helper.mjs';
import '../../../_chunks/dep-c3ca6d75.mjs';
import '../../../_chunks/dep-ab0f22e3.mjs';
import '../../../_chunks/dep-d45d8d10.mjs';
import '../../../input/useInputEventHandler.mjs';
import '../../../input/useInputWidth.mjs';
import '../../../input/input-group.mjs';
import '../../../input/input-group-props.mjs';
import '../../../input/type.mjs';
import '../../../loading/index.mjs';
import '../../../_chunks/dep-a210005e.mjs';
import '../../../_chunks/dep-ff970c88.mjs';
import '../../../_chunks/dep-945356ed.mjs';
import '../../../_chunks/dep-0554a988.mjs';
import '../../../loading/plugin.mjs';
import '../../../loading/loading.mjs';
import '../../../loading/icon/gradient.mjs';
import '../../../_common/js/loading/circle-adapter.mjs';
import '../../../loading/props.mjs';
import '../../../hooks/useTeleport.mjs';
import '../../../loading/type.mjs';
import '../../../select-input/useMultiple.mjs';
import '../../../tag-input/index.mjs';
import '../../../tag-input/tag-input.mjs';
import '../../../tag-input/props.mjs';
import '../../../tag-input/hooks/useTagScroll.mjs';
import '../../../tag-input/useTagList.mjs';
import '../../../tag/index.mjs';
import '../../../tag/tag.mjs';
import '../../../_chunks/dep-4c53eba6.mjs';
import '../../../tag/props.mjs';
import '../../../tag/check-tag.mjs';
import '../../../tag/check-tag-props.mjs';
import '../../../_common/js/common.mjs';
import '../../../tag/check-tag-group.mjs';
import '../../../tag/check-tag-group-props.mjs';
import '../../../tag/type.mjs';
import '../../../tag-input/hooks/useHover.mjs';
import '../../../tag-input/hooks/useDragSorter.mjs';
import '../../../tag-input/type.mjs';
import '../../../select-input/useOverlayInnerStyle.mjs';
import '../../../select-input/type.mjs';
import '../../../select/select-panel.mjs';
import '../../../_chunks/dep-3aafefae.mjs';
import '../../../_chunks/dep-385f320b.mjs';
import '../../../select/option.mjs';
import '../../../select/option-props.mjs';
import '../../../checkbox/index.mjs';
import '../../../checkbox/checkbox.mjs';
import '../../../checkbox/props.mjs';
import '../../../checkbox/constants.mjs';
import '../../../checkbox/hooks/useCheckboxLazyLoad.mjs';
import '../../../_common/js/utils/observe.mjs';
import '../../../checkbox/hooks/useKeyboardEvent.mjs';
import '../../../checkbox/group.mjs';
import '../../../checkbox/checkbox-group-props.mjs';
import '../../../checkbox/type.mjs';
import '../../../select/helper.mjs';
import '../../../select/optionGroup.mjs';
import '../../../select/option-group-props.mjs';
import '../../../select/props.mjs';
import '../../../select/hooks/usePanelVirtualScroll.mjs';
import '../../../select/hooks/useSelectOptions.mjs';
import '../../../select/hooks/useKeyboardControl.mjs';
import '../../../select/type.mjs';
import '../../../_chunks/dep-cf4a2c01.mjs';
import '../../utils/index.mjs';
import '../../../_common/js/color-picker/index.mjs';
import '../../../_common/js/color-picker/cmyk.mjs';
import '../../../_common/js/color-picker/color.mjs';
import '../../../_chunks/dep-2c57cd88.mjs';
import '../../../_common/js/color-picker/gradient.mjs';
import '../../../_common/js/color-picker/draggable.mjs';
import '../../../input-number/index.mjs';
import '../../../input-number/input-number.mjs';
import '../../../button/index.mjs';
import '../../../button/button.mjs';
import '../../../button/props.mjs';
import '../../../button/type.mjs';
import '../../../input-number/props.mjs';
import '../../../input-number/hooks/useInputNumber.mjs';
import '../../../_common/js/input-number/number.mjs';
import '../../../_common/js/input-number/large-number.mjs';
import '../../../input-number/type.mjs';
import './config.mjs';

var createCompounder = _createCompounder;

/**
 * Converts `string`, as space separated words, to upper case.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category String
 * @param {string} [string=''] The string to convert.
 * @returns {string} Returns the upper cased string.
 * @example
 *
 * _.upperCase('--foo-bar');
 * // => 'FOO BAR'
 *
 * _.upperCase('fooBar');
 * // => 'FOO BAR'
 *
 * _.upperCase('__foo_bar__');
 * // => 'FOO BAR'
 */
var upperCase = createCompounder(function (result, word, index) {
  return result + (index ? ' ' : '') + word.toUpperCase();
});
var upperCase_1 = upperCase;

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var FormatPanel = defineComponent({
  name: "FormatPanel",
  inheritAttrs: false,
  props: _objectSpread(_objectSpread({}, props), {}, {
    color: {
      type: Object
    },
    onModeChange: {
      type: Function,
      "default": function _default() {
        return function () {};
      }
    },
    onInputChange: {
      type: Function,
      "default": function _default() {
        return function () {};
      }
    }
  }),
  setup: function setup(props2) {
    var baseClassName = useBaseClassName();
    var formatModel = ref(props2.format);
    watch(function () {
      return [props2.format];
    }, function () {
      return formatModel.value = props2.format;
    });
    var handleModeChange = function handleModeChange(v) {
      formatModel.value = v;
      props2.onModeChange(v);
    };
    return {
      formatModel: formatModel,
      baseClassName: baseClassName,
      handleModeChange: handleModeChange
    };
  },
  render: function render() {
    var _this = this;
    var _slot;
    var formats = _toConsumableArray(FORMATS);
    var baseClassName = this.baseClassName,
      handleModeChange = this.handleModeChange;
    var newProps = _objectSpread(_objectSpread({}, this.$props), {}, {
      format: this.formatModel
    });
    var selectInputProps = _objectSpread({}, this.selectInputProps || {});
    return createVNode("div", {
      "class": "".concat(baseClassName, "__format")
    }, [createVNode("div", {
      "class": "".concat(baseClassName, "__format--item")
    }, [createVNode(Select, {
      "size": "small",
      "class": "".concat(baseClassName, "__format-mode-select"),
      "selectInputProps": _objectSpread({}, selectInputProps),
      "popupProps": _objectSpread({
        overlayClassName: "".concat(baseClassName, "__select-options")
      }, selectInputProps.popupProps),
      "modelValue": _this.formatModel,
      "onUpdate:modelValue": function onUpdateModelValue($event) {
        return _this.formatModel = $event;
      },
      "onChange": handleModeChange
    }, _isSlot(_slot = formats.map(function (item) {
      return createVNode(Option, {
        "key": item,
        "value": item,
        "label": upperCase_1(item),
        "style": {
          fontSize: "12px"
        }
      }, null);
    })) ? _slot : {
      "default": function _default() {
        return [_slot];
      }
    })]), createVNode("div", {
      "class": "".concat(baseClassName, "__format--item")
    }, [createVNode(FormatInputs, newProps, null)])]);
  }
});

export { FormatPanel as default };
