// tdesign-vue-next/es/color-picker/panel/format/inputs.mjs
/**
 * tdesign v1.10.2
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _toConsumableArray } from '../../../_chunks/dep-5ce31aa3.mjs';
import { _ as _defineProperty } from '../../../_chunks/dep-5597cc76.mjs';
import { defineComponent, computed, reactive, watch, createVNode, mergeProps } from 'vue';
import { t as throttle_1 } from '../../../_chunks/dep-cf4a2c01.mjs';
import props from '../../props.mjs';
import '../../utils/index.mjs';
import { Input } from '../../../input/index.mjs';
import { InputNumber } from '../../../input-number/index.mjs';
import { FORMAT_INPUT_CONFIG } from './config.mjs';
import { Color } from '../../../_common/js/color-picker/color.mjs';
import '../../../_chunks/dep-9213781b.mjs';
import '../../../_chunks/dep-670ed189.mjs';
import '../../../_chunks/dep-95eb038b.mjs';
import '../../../_chunks/dep-8b2e46b6.mjs';
import '../../../_chunks/dep-48219979.mjs';
import '../../../_chunks/dep-2b538834.mjs';
import '../../../_chunks/dep-19284555.mjs';
import '../../../_chunks/dep-6eed8561.mjs';
import '../../../_chunks/dep-b02becad.mjs';
import '../../../_chunks/dep-5b72198e.mjs';
import '../../../_common/js/color-picker/index.mjs';
import '../../../_common/js/color-picker/cmyk.mjs';
import '../../../_common/js/color-picker/draggable.mjs';
import '../../../_chunks/dep-2c57cd88.mjs';
import '../../../_common/js/color-picker/gradient.mjs';
import '../../../_chunks/dep-6e744142.mjs';
import '../../../_chunks/dep-0e4e2bcf.mjs';
import '../../../_chunks/dep-9a145a1d.mjs';
import '../../../_chunks/dep-c3ca6d75.mjs';
import '../../../_chunks/dep-4c53eba6.mjs';
import '../../../_chunks/dep-97ca0f39.mjs';
import '../../../input/input.mjs';
import '../../../_chunks/dep-209a9bdd.mjs';
import 'tdesign-icons-vue-next';
import '../../../input/props.mjs';
import '../../../hooks/useDisabled.mjs';
import '../../../_chunks/dep-47d8cf00.mjs';
import '../../../hooks/useReadonly.mjs';
import '../../../hooks/useConfig.mjs';
import '../../../config-provider/useConfig.mjs';
import '../../../_chunks/dep-6f372498.mjs';
import '../../../_chunks/dep-95651bdf.mjs';
import '../../../_chunks/dep-37133d6d.mjs';
import '../../../_chunks/dep-18e89938.mjs';
import '../../../_chunks/dep-50ae0ec6.mjs';
import '../../../_chunks/dep-c32ef8ef.mjs';
import '../../../_chunks/dep-eae00e6e.mjs';
import '../../../_chunks/dep-40d5a6ec.mjs';
import '../../../_chunks/dep-564c533a.mjs';
import '../../../_chunks/dep-1b4b86fa.mjs';
import '../../../_chunks/dep-8eb27cbe.mjs';
import '../../../_chunks/dep-257497ac.mjs';
import '../../../_chunks/dep-f5c04f3c.mjs';
import '../../../_chunks/dep-6623c796.mjs';
import '../../../_chunks/dep-70cd6787.mjs';
import '../../../_chunks/dep-31a95ba9.mjs';
import '../../../_chunks/dep-c4e7b9d7.mjs';
import '../../../_chunks/dep-3f5bea17.mjs';
import '../../../_chunks/dep-d22c9e6d.mjs';
import '../../../_chunks/dep-a27f8ab7.mjs';
import '../../../_chunks/dep-39e58764.mjs';
import '../../../_chunks/dep-a2bb880f.mjs';
import '../../../_chunks/dep-bf0849ad.mjs';
import '../../../_chunks/dep-4ef46503.mjs';
import '../../../_common/js/global-config/default-config.mjs';
import '../../../_common/js/global-config/locale/zh_CN.mjs';
import '../../../_chunks/dep-257af9e6.mjs';
import '../../../_chunks/dep-c34f015e.mjs';
import '../../../_chunks/dep-7e9fb780.mjs';
import '../../../config-provider/type.mjs';
import '../../../hooks/useGlobalIcon.mjs';
import '../../../hooks/tnode.mjs';
import '../../../_chunks/dep-8d9b7c94.mjs';
import '../../../_chunks/dep-c3300d46.mjs';
import '../../../_chunks/dep-399b9c3e.mjs';
import '../../../_chunks/dep-684927fd.mjs';
import '../../../_chunks/dep-82139152.mjs';
import '../../../_chunks/dep-b88cf96a.mjs';
import '../../../utils/render-tnode.mjs';
import '../../../_chunks/dep-dfc325a2.mjs';
import '../../../input/useInput.mjs';
import '../../../form/const.mjs';
import '../../../hooks/index.mjs';
import '../../../hooks/icon.mjs';
import '../../../hooks/slot.mjs';
import '../../../hooks/useCommonClassName.mjs';
import '../../../hooks/useDefaultValue.mjs';
import '../../../hooks/useDestroyOnClose.mjs';
import '../../../hooks/useKeepAnimation.mjs';
import '../../../hooks/useRipple.mjs';
import '../../../utils/set-style.mjs';
import '../../../hooks/useVirtualScroll.mjs';
import '../../../hooks/useVModel.mjs';
import '../../../hooks/useImagePreviewUrl.mjs';
import '../../../_common/js/upload/utils.mjs';
import '../../../_common/js/log/log.mjs';
import '../../../input/useLengthLimit.mjs';
import '../../../_common/js/log/index.mjs';
import '../../../_common/js/utils/helper.mjs';
import '../../../_chunks/dep-ab0f22e3.mjs';
import '../../../_chunks/dep-d45d8d10.mjs';
import '../../../input/useInputEventHandler.mjs';
import '../../../input/useInputWidth.mjs';
import '../../../hooks/useResizeObserver.mjs';
import '../../../input/input-group.mjs';
import '../../../input/input-group-props.mjs';
import '../../../utils/withInstall.mjs';
import './style/css.mjs';
import '../../../input/type.mjs';
import '../../../input-number/input-number.mjs';
import '../../../button/index.mjs';
import '../../../button/button.mjs';
import '../../../loading/index.mjs';
import '../../../_chunks/dep-a210005e.mjs';
import '../../../_chunks/dep-ff970c88.mjs';
import '../../../_chunks/dep-644e70b3.mjs';
import '../../../_chunks/dep-945356ed.mjs';
import '../../../_chunks/dep-e9c00cc1.mjs';
import '../../../_chunks/dep-688d94c5.mjs';
import '../../../_chunks/dep-a9ad4d99.mjs';
import '../../../_chunks/dep-0554a988.mjs';
import '../../../loading/plugin.mjs';
import '../../../loading/loading.mjs';
import '../../../loading/icon/gradient.mjs';
import '../../../_common/js/loading/circle-adapter.mjs';
import '../../../_common/js/utils/set-style.mjs';
import '../../../utils/dom.mjs';
import '../../../utils/easing.mjs';
import '../../../loading/props.mjs';
import '../../../hooks/useTeleport.mjs';
import '../../../loading/type.mjs';
import '../../../button/props.mjs';
import '../../../button/type.mjs';
import '../../../input-number/props.mjs';
import '../../../input-number/hooks/useInputNumber.mjs';
import '../../../_common/js/input-number/number.mjs';
import '../../../_common/js/input-number/large-number.mjs';
import '../../../input-number/type.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var FormatInputs = defineComponent({
  name: "FormatInputs",
  inheritAttrs: false,
  props: _objectSpread(_objectSpread({}, props), {}, {
    color: {
      type: Object
    },
    onInputChange: {
      type: Function,
      "default": function _default() {
        return function () {};
      }
    }
  }),
  setup: function setup(props2) {
    var inputConfigs = computed(function () {
      var configs = _toConsumableArray(FORMAT_INPUT_CONFIG[props2.format]);
      if (props2.enableAlpha) {
        configs.push({
          type: "inputNumber",
          key: "a",
          min: 0,
          max: 100,
          format: function format(value) {
            return "".concat(value, "%");
          },
          flex: 1.15
        });
      }
      return configs;
    });
    var modelValue = reactive({});
    var lastModelValue = reactive({});
    var getFormatColorMap = function getFormatColorMap(type) {
      var color = props2.color;
      if (type === "encode") {
        return {
          HSV: color.getHsva(),
          HSL: color.getHsla(),
          RGB: color.getRgba(),
          CMYK: color.getCmyk(),
          CSS: {
            css: color.css
          },
          HEX: {
            hex: color.hex
          }
        };
      }
      return {
        HSV: Color.object2color(modelValue, "HSV"),
        HSL: Color.object2color(modelValue, "HSL"),
        RGB: Color.object2color(modelValue, "RGB"),
        CMYK: Color.object2color(modelValue, "CMYK"),
        CSS: modelValue.css,
        HEX: modelValue.hex
      };
    };
    var updateModelValue = function updateModelValue() {
      var format = props2.format,
        color = props2.color;
      var values = getFormatColorMap("encode")[format];
      values.a = Math.round(color.alpha * 100);
      Object.keys(values).forEach(function (key) {
        modelValue[key] = values[key];
        lastModelValue[key] = values[key];
      });
    };
    updateModelValue();
    var throttleUpdate = throttle_1(updateModelValue, 100);
    watch(function () {
      var _props2$color = props2.color,
        saturation = _props2$color.saturation,
        hue = _props2$color.hue,
        value = _props2$color.value,
        alpha = _props2$color.alpha,
        css = _props2$color.css;
      return [saturation, hue, value, alpha, css, props2.format];
    }, throttleUpdate);
    var handleChange = function handleChange(key, v) {
      if (v === lastModelValue[key]) {
        return;
      }
      var value = getFormatColorMap("decode")[props2.format];
      props2.onInputChange(value, modelValue.a / 100, key, v);
    };
    return {
      modelValue: modelValue,
      inputConfigs: inputConfigs,
      handleChange: handleChange
    };
  },
  render: function render() {
    var _this = this;
    var inputProps = _objectSpread({}, this.inputProps || {});
    return createVNode("div", {
      "class": "input-group"
    }, [this.inputConfigs.map(function (config) {
      return createVNode("div", {
        "class": "input-group__item",
        "key": config.key,
        "style": {
          flex: config.flex || 1
        }
      }, [config.type === "input" ? createVNode(Input, mergeProps(inputProps, {
        "align": "center",
        "size": "small",
        "disabled": _this.disabled,
        "modelValue": _this.modelValue[config.key],
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return _this.modelValue[config.key] = $event;
        },
        "maxlength": _this.format === "HEX" ? 9 : void 0,
        "title": _this.modelValue[config.key],
        "onBlur": function onBlur(v) {
          return _this.handleChange(config.key, v);
        },
        "onEnter": function onEnter(v) {
          return _this.handleChange(config.key, v);
        }
      }), null) : createVNode(InputNumber, mergeProps(inputProps, {
        "align": "center",
        "size": "small",
        "disabled": _this.disabled,
        "modelValue": _this.modelValue[config.key],
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return _this.modelValue[config.key] = $event;
        },
        "title": _this.modelValue[config.key],
        "min": config.min,
        "max": config.max,
        "step": 1,
        "format": config.format,
        "theme": "normal",
        "onBlur": function onBlur(v) {
          return _this.handleChange(config.key, v);
        },
        "onEnter": function onEnter(v) {
          return _this.handleChange(config.key, v);
        }
      }), null)]);
    })]);
  }
});

export { FormatInputs as default };
